### PR TITLE
DbAPiHook: Don't log a warning message if placeholder is None and make sure warning message is formatted correctly

### LINKED ISSUE #39678: GenericTransfer - Warning: "Placeholder defined in Connection..."
### Apache Airflow version

2.9.1

### If "Other Airflow 2 version" selected, which one?

_No response_

### What happened?

We use GenericTransfer task to copy data from one DB to another. Task is successful but we get a warning in logs:
`[2024-05-17, 10:08:43 UTC] {sql.py:187} WARNING - Placeholder defined in Connection 'None' is not listed in 'DEFAULT_SQL_PLACEHOLDERS' and got ignored. Falling back to the default placeholder '%s'.`

I checked the Connections in the Admin and everything looks alright.

Screenshot 2024-05-17 at 12 22 33

### What you think should happen instead?

No warning log

### How to reproduce

create and run GenericTransfer task:

```python
copy_task = GenericTransfer(
	task_id = 'copy_task',
	dag = dag,
	source_conn_id = 'source-db',
	destination_conn_id = 'bi-database',
	sql = '''
		SELECT
			"date"
		FROM "TableName"
	''',
	destination_table = '"TableName"',
	preoperator = '''
		DELETE FROM "TableName"
		WHERE
			1 = 1
	'''
)
```

### Operating System

Debian GNU/Linux 12 (bookworm)

### Versions of Apache Airflow Providers

apache-airflow-providers-amazon==8.20.0
apache-airflow-providers-celery==3.6.2
apache-airflow-providers-cncf-kubernetes==8.1.1
apache-airflow-providers-common-io==1.3.1
apache-airflow-providers-common-sql==1.12.0
apache-airflow-providers-fab==1.0.4
apache-airflow-providers-ftp==3.8.0
apache-airflow-providers-http==4.10.1
apache-airflow-providers-imap==3.5.0
apache-airflow-providers-postgres==5.10.2
apache-airflow-providers-smtp==1.6.1
apache-airflow-providers-sqlite==3.7.1

### Deployment

Other Docker-based deployment

### Deployment details

Airflow 2.9.1, on-premise Kubernetes cluster 1.26.5, python 3.10.12, Celery executor, Postgres 13.13

### Anything else?

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by dabla**
closes: #39678

This PR fixes an invalid warning message being logged when placeholder is None (and thus not defined in the connection extra field) and also makes sure that when the placeholder isn't valid, a correct warning message is being logged.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: tests/providers/common/sql/hooks/test_dbapi.py [hunk 1]
```diff
@@ -421,6 +421,56 @@ def test_get_uri_without_auth_and_empty_host(self):
...
+        )
+
+        assert self.db_hook.placeholder == "%s"
+
+        self.db_hook.log.warning.assert_called_once_with(
+            "Placeholder defined in Connection '%s' is not listed in 'DEFAULT_SQL_PLACEHOLDERS' "
+            "and got ignored. Falling back to the default placeholder '%s'.",
+            "test_conn_id",
+            self.db_hook._placeholder,
+        )
```
Comment (Taragolis) [inline tests/providers/common/sql/hooks/test_dbapi.py]: https://docs.pytest.org/en/stable/how-to/logging.html#caplog-fixture

Comment (dabla) [reply tests/providers/common/sql/hooks/test_dbapi.py]: Problem with this approach, as I already tried, is that I'm unable to capture the debug statements.  Also previous tests where also already done that way, not that I don't want to do it, but I'm afraid the PR will take more time to try to figure this out.  Info statements get captured, and yes I already tried with this:
```
    def test_insert_rows_as_generator_supports_executemany(self, caplog):
        table = "table"
        rows = [("What's",), ("up",), ("world",)]

        with caplog.at_level(logging.DEBUG):
            self.db_hook.supports_executemany = True
            self.db_hook.insert_rows(table, iter(rows))

        assert self.conn.close.call_count == 1
        assert self.cur.close.call_count == 1
        assert self.conn.commit.call_count == 2

        sql = f"INSERT INTO {table}  VALUES (%s)"

        assert any(f"Generated sql: {sql}" in message for message in caplog.messages) # this one isn't captured
        assert any(f"Loaded 3 rows into {table} so far" in message for message in caplog.messages) # works
        assert any(
            f"Done loading. Loaded a total of 3 rows into {table}" in message # works
            for message in caplog.messages
        )

        self.cur.executemany.assert_any_call(sql, rows)
```

Comment (dabla) [reply tests/providers/common/sql/hooks/test_dbapi.py]: Never mind found a solution

Comment (Taragolis) [reply tests/providers/common/sql/hooks/test_dbapi.py]: Yeah I saw that previous test captured by this kind of the.
But better if possible capture actual logging.

BTW, you could also use, but there is no preferable way how to extract messages from captured logs

```python
assert "something" in caplog.text
```

**Review by Taragolis — APPROVED** [review]

**Review by dirrao — APPROVED** [review]

**Pull request merged**